src/db_cache.rs

```diff
@@ -1,188 +1,171 @@
+///         self.inner.len() as u64
+///     }
+/// }
+///
+/// #[::tokio::main]
+/// async fn main() {
+///     let object_store = Arc::new(LocalFileSystem::new());
+///     let options = DbOptions {
+///         block_cache: Some(Arc::new(MyCache::new())),
+///         ..Default::default()
+///     };
+///     let db = Db::open_with_opts("path/to/db".into(), options, object_store).await;
+/// }
+/// ```
 #[async_trait]
-pub(crate) trait DbCache: Send + Sync + 'static {
-    async fn get(&self, key: (SsTableId, u64)) -> CachedBlockOption;
-    async fn insert(&self, key: (SsTableId, u64), value: CachedBlock);
+pub trait DbCache: Send + Sync {
+    async fn get(&self, key: CachedKey) -> Option<CachedEntry>;
+    async fn insert(&self, key: CachedKey, value: CachedEntry);
     #[allow(dead_code)]
-    async fn remove(&self, key: (SsTableId, u64));
+    async fn remove(&self, key: CachedKey);
     #[allow(dead_code)]
     fn entry_count(&self) -> u64;
 }
 
-pub(crate) struct MokaCache {
-    inner: moka::future::Cache<(SsTableId, u64), CachedBlock>,
-}
-
-impl MokaCache {
-    pub fn new(options: DbCacheOptions) -> Self {
-        let mut builder = moka::future::Cache::builder()
-            .weigher(move |_, _| options.cached_block_size)
-            .max_capacity(options.max_capacity);
-
-        if let Some(ttl) = options.time_to_live {
-            builder = builder.time_to_live(ttl);
-        }
-
-        if let Some(tti) = options.time_to_idle {
-            builder = builder.time_to_idle(tti);
-        }
-
-        let cache = builder.build();
-
-        Self { inner: cache }
+/// A key used to identify a cached entry.
+///
+/// The key is a tuple of an SSTable ID and a block ID.
+/// The tuple is private to this module, so the implementation details
+/// of the cache are not exposed publicly.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct CachedKey(SsTableId, u64);
+
+impl From<(SsTableId, u64)> for CachedKey {
+    fn from((sst_id, block_id): (SsTableId, u64)) -> Self {
+        Self(sst_id, block_id)
     }
 }
 
-#[async_trait]
-impl DbCache for MokaCache {
-    async fn get(&self, key: (SsTableId, u64)) -> CachedBlockOption {
-        CachedBlockOption::Moka(self.inner.get(&key).await)
-    }
-
-    async fn insert(&self, key: (SsTableId, u64), value: CachedBlock) {
-        self.inner.insert(key, value).await;
-    }
-
-    async fn remove(&self, key: (SsTableId, u64)) {
-        self.inner.remove(&key).await;
-    }
-
-    fn entry_count(&self) -> u64 {
-        self.inner.entry_count()
-    }
+#[derive(Clone)]
+enum CachedItem {
+    Block(Arc<Block>),
+    SsTableIndex(Arc<SsTableIndexOwned>),
+    BloomFilter(Arc<BloomFilter>),
 }
 
-pub(crate) struct FoyerCache {
-    inner: foyer::Cache<(SsTableId, u64), CachedBlock>,
+/// A cached entry stored in the cache.
+///
+/// The entry stores data in an internal enum that represents the type of cached item.
+/// The enum is private to this module, so the implementation details
```

Comment:
nit: I'd rephrase to "the internal types of the entries that are stored in the cache"

Reply:
updated